### PR TITLE
feat: improved the README.md for AnyMaths; fixed AnyMathsAdapter code; provided example training and dataset preparation; added eval results and sample eval code

### PR DESCRIPTION
**Pull request by egmaminta**
#### Summary
This PR improves the `AnyMaths` example end-to-end by fixing the adapter code (I found a bug that only outputs 1 Pareto val set), adding dataset preparation utilities, providing a runnable sample training script, and updating the `AnyMaths` README with clear usage guidance.

#### Changes
- Docs
    - Updated `README.md` with setup, configuration, and run instructions.
- Adapter
    - Fixed `AnyMaths` adapter to correctly handle multi-item evaluation (no more assertion error when valset > 1).
    - Ensured batched/iterative evaluation returns consistent metrics and aggregates correctly.
- Dataset preparation
    - Added dataset loading and splitting logic supporting:
        - `"openai/gsm8k"` (train/test) with deterministic shuffling and derived solutions.
        - `"MathArena/aime_2025"` with aligned train/test preprocessing.
    - Provides train/val split from train; expands a small test set for demo runs.
- Sample training
    - Added `train_anymaths.py`:
        - CLI args for model, API base, budget, reflection LM, minibatch size, and seed.
        - Uses `gepa.optimize` with `AnyMathsAdapter`.
        - LiteLLM-based reflection LM hook.
        - Small default subset sizes for quick demos.
- Benchmark results
    - Added `openai/gsm8k` - `test` split results of `ollama/qwen3:4b`. (Before GEPA: 18% accuracy; After GEPA: **23%** accuracy; Budget used: 5).
    - Added a sample eval script for `openai/gsm8k` (see `src/gepa/examples/anymaths-bench/eval_default.py`).

#### Why
- The adapter previously triggered an assertion when `valset > 1`. This is non-ideal for real validation and made experiments brittle.
- Providing dataset prep and a runnable script lowers the barrier to trying AnyMaths with GEPA.

#### How to use
- Start an Ollama server (or compatible endpoint) and ensure the requested models are available.
- Run the training script (example):
    - Module: `python -m gepa.examples.anymaths-bench.train_anymaths --model_name ollama/qwen3:4b --api_base http://localhost:11434 --anymaths_dset_name openai/gsm8k --budget 500`
- Adjust `--reflection_lm`, `--reflection_minibatch_size`, and `--seed` as needed.

#### Notes
- Defaults limit train/val/test to 10 items for fast demos; expand as desired.
- No breaking API changes expected outside the `AnyMaths` example path.

Checklist

- [x] Docs updated
- [x] Adapter supports multi-item validation
- [x] Dataset prep utilities included
- [x] Sample training script added and runnable
- [x] Added experiment results on `ollama/qwen3:4b`
- [x] Added observation on the optimal prompt

### CODE REVIEW

File: src/gepa/adapters/anymaths_adapter/README.md [hunk 1]
```diff
@@ -1,7 +1,195 @@
...
+    "answer": ...
+}
+```
+- `question` must be a string/text.
+- `solution` must be a string/text.
+- `answer` must be a purely numerical, no other text or units associated.
+
+It is best to upload your custom dataset to the Hugging Face dataset hub to fully utilize `datasets.load_dataset`.
+
+### Ollama-only
```
Comment (LakshyAAAgrawal) [inline src/gepa/adapters/anymaths_adapter/README.md]: Question: Why can't this work directly with any of the other models supported by litellm? What is specific to Ollama here?

IMO, we can highlight that this adapter can work even locally without expensive GPUs, using Ollama.

**Comment by LakshyAAAgrawal** [comment]
Could you also go through [CONTRIBUTING.md](CONTRIBUTING.md) to setup the linter, and address the above failing check?

File: src/gepa/adapters/anymaths_adapter/README.md (hunk 1 above)
Comment (LakshyAAAgrawal) [reply src/gepa/adapters/anymaths_adapter/README.md]: This does not address my concern: The concern is, let's say I have access to an OpenAI API Key, or OpenRouter API Key, why can't I use this adapter with them?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 default_language_version:
-  python: python3.11
+  python: python3.10
```
Comment (LakshyAAAgrawal) [inline .pre-commit-config.yaml]: Please undo this change.

File: src/gepa/adapters/anymaths_adapter/README.md [hunk 3]
```diff
@@ -1,7 +1,195 @@
 # GEPA for Math Problem Solving
-AnyMaths Adapter is a GEPAAdapter for any dataset that contains mathematical word problems of varying complexity and structure. It is designed to handle a wide range of mathematical tasks, including arithmetic, algebra, and more.
+**`AnyMaths`** Adapter is a GEPAAdapter for any dataset that contains mathematical word problems of varying complexity and structure. It is designed to handle a wide range of mathematical tasks, including arithmetic, algebra, and more.
 
-Note: Ollama must be installed and configured to use this adapter.
+Note: `Ollama` must be installed and configured to use this adapter. Instructions to download Ollama can be found in [this link](https://ollama.com/download). I recommend the Linux installation.
```
Comment (LakshyAAAgrawal) [inline src/gepa/adapters/anymaths_adapter/README.md]: Sorry for flagging this again, but I still don't understand why is Ollama a requirement? I understand that it is a feature, that even without GPUs, one can run this adapter. However, I think the adapter can be used as-is even without Ollama by pointing to any external API. Please correct me if I am missing something.

File: src/gepa/adapters/anymaths_adapter/README.md [hunk 4]
```diff
@@ -1,7 +1,195 @@
...
+```
+You are an AI assistant that solves mathematical word problems. You will be given a question and you need to provide a step-by-step solution to the problem. Finally, you will provide the answer to the question.
+
+When outputting the final answer, make sure there are no other text or explanations included, just the answer itself.
+
+The following fields are what you need to include in your response:
+- final_answer: The final answer to the question.
+- solution_pad: The step-by-step solution to the problem.
+```
+- Final prompt (after 4 iterations):
```
Comment (LakshyAAAgrawal) [inline src/gepa/adapters/anymaths_adapter/README.md]: Can you clarify what is the meaning of iterations here?

File: src/gepa/examples/anymaths-bench/train_anymaths.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        seed_candidate={"instruction_prompt": seed_instruction},
+        trainset=trainset,
+        valset=valset,
+        adapter=AnyMathsAdapter(model=adapter_model, api_base=api_base, max_litellm_workers=max_litellm_workers),
+        reflection_lm=reflection_lm,
+        reflection_minibatch_size=reflection_minibatch_size,
+        perfect_score=1,
+        skip_perfect_score=False,
+        use_wandb=False,
+        num_iters=budget,
```
Comment (LakshyAAAgrawal) [inline src/gepa/examples/anymaths-bench/train_anymaths.py]: I just merged a PR removing the num_iters argument. Can you convert this into num rollouts instead?

**Review by LakshyAAAgrawal — CHANGES_REQUESTED** [review]

**Review by LakshyAAAgrawal — APPROVED** [review]
